src/precice/impl/ParticipantImpl.cpp

```diff
@@ -392,6 +392,15 @@ void ParticipantImpl::advance(
 
   advanceCouplingScheme();
 
+  if (_couplingScheme->isTimeWindowComplete()) {
+    for (auto &context : _accessor->readDataContexts()) {
+      context.resetInitialGuesses();
+    }
+    for (auto &context : _accessor->writeDataContexts()) {
+      context.resetInitialGuesses();
+    }
+  }
```

Comment:
This should only be the case if all coupingschemes have converged, right?

I think this currently always resets the initial guess  as soon as we reach the end of the time window, meaning after each iteration.

Pretty counterproductive

Reply:
> This should only be the case if all coupingschemes have converged, right?

Yes, we only want to rest if `isTimeWindowComplete()`, but we could also call this within `mapRead/WrittenData`.

> Is this the correct location for all coupling schemes?

I am mainly concerned with serial vs parallel coupling (and combining this with WR). Maybe this is the correct location, but I have not thought about this at length. 
Maybe @BenjaminRodenberg could help?

Reply:
Hard to say, because I am not that familiar with what `resetInitialGuesses()` is doing. A dedicated test for this piece of code would be very helpful, because I guess this is something that might easily break in the future (maybe already when I finalize the WR).

Reply:
Ok, `_couplingScheme->isTimeWindowComplete` is true when the coupling reaches the end of the time window.
For compositional cplschemes, all schemes need to return true.

Reply:
I still think that this is the wrong location. It should happen after the read mapping as the data mapped there still belongs to the converged time step. Only afterwards, once we leave `advance()`, the solvers switch to the next time step (by not reloading a checkpoint).

We need to catch this behavior with a test to safeguard against future changes.

We cannot, however, move this call into `mapRead/WrittenData` as the `mapWrittenData` does not yet know whether the time window is complete.

Reply:
Mmh, true, complicated. Needs to be a whitebox test. We could check the stored initial guess of the `DataContext`, but this would neither allow us to check for the correct location of the reset within one advance call.
If too complicated and too little benefit, let's drop.

Reply:
I'll open an issue. Maybe we can figure something out in the future.

Reply:
https://github.com/precice/precice/issues/1688